### PR TITLE
C++ API: torch::nn::Softmax2d

### PR DESCRIPTION
**Pull request by nuka137**
Add torch::nn::Softmax2d module support for the C++ API.
Softmax2d only supports module in Python API, so this PR adds only module support as well.

This PR is WIP because it uses the function in #27446 .
After #27446 is merged, I will remove WIP.

Related Issue: #25883 

Reviewer: @yf225

### CODE REVIEW

**Comment by yf225** [comment]
Thanks a lot for the great work! We will need some fixes as suggested in https://github.com/pytorch/pytorch/pull/27446.

**Comment by nuka137** [comment]
@yf225 

#27446 is merged, so I removed WIP.
Could you review this PR again?

File: torch/csrc/api/include/torch/nn/modules/activation.h [hunk 1]
```diff
@@ -180,6 +180,25 @@ class TORCH_API LogSoftmaxImpl : public torch::nn::Cloneable<LogSoftmaxImpl> {
...
 TORCH_MODULE(LogSoftmax);
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Softmax2d ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+/// Applies the Softmax2d function element-wise.
+/// See https://pytorch.org/docs/master/nn.html#torch.nn.Softmax2d to learn
+/// about the exact behavior of this module.
+class TORCH_API Softmax2dImpl : public torch::nn::Cloneable<Softmax2dImpl> {
+ public:
+  Softmax2dImpl();
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/modules/activation.h]: We can remove the default constructor because it will be created by default :D

**Review by yf225 — COMMENTED** [review]
Thanks so much for the awesome work @nuka137! I left a very minor comment, and we should be able to merge it very soon :D

File: torch/csrc/api/src/nn/modules/activation.cpp [hunk 2]
```diff
@@ -146,6 +146,21 @@ Tensor LogSoftmaxImpl::forward(const Tensor& input) {
 
 // ============================================================================
 
+Softmax2dImpl::Softmax2dImpl() {}
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/activation.cpp]: We can remove the default constructor here :D

File: torch/csrc/api/include/torch/nn/modules/activation.h (hunk 1 above)
Comment (nuka137) [reply torch/csrc/api/include/torch/nn/modules/activation.h]: Fixed it

File: torch/csrc/api/src/nn/modules/activation.cpp (hunk 2 above)
Comment (nuka137) [reply torch/csrc/api/src/nn/modules/activation.cpp]: OK, I fixed it.

**Comment by nuka137** [comment]
@yf225 

Thanks for reviewing.
I fixed all issues you mentioned.
Could you review again?

**Review by yf225 — APPROVED** [review]
Thanks so much for the awesome work @nuka137!

**Pull request closed without merging**